app/controllers/application_controller/buttons.rb

```diff
@@ -147,7 +177,58 @@ def automate_button_field_changed
         page.replace("form_role_visibility", :partial => "layouts/role_visibility", :locals => {:rec_id => (@custom_button.id || "new").to_s, :action => "automate_button_field_changed"})
       end
       unless params[:target_class]
-        @changed = session[:changed] = (@edit[:new] != @edit[:current])
+        @changed = session[:changed] = false
+
+        # This block checks if any of the fields have changed
+        # Broken into different branches for specific fields
+        # If no fields are changed it will not set @changed or session[:changed] to true
+        @edit[:new].each_key do |key|
+          if @edit[:new][key] != @edit[:current][key]
+            if @edit[:new][key] == "" && @edit[:current][key].nil? # check empty string / nil case
+              next
+            elsif @edit[:new][key] == @edit[:current][key].to_s # check string / integer case
+              next
+            elsif key == :roles # check role values
+              if (@edit[:new][key] == ["_ALL_"] || @edit[:new][key] == []) && (@edit[:current][key].nil? || @edit[:current][key] == ["all"] || @edit[:current][key] == ["_ALL_"]) # if visibility is set to "To All"
+                next
+              elsif @edit[:new][key].sort == @edit[:current][key].sort # check if new roles array and current roles array are equal after sorting the ids
+                next
+              else # if new roles array and current roles array are not equal
+                @changed = session[:changed] = true
+                break
+
+              end
+            elsif key == :attrs # check attribute values
+              @edit[:new][key].each_with_index do |_item, index|
+                if (@edit[:new][:attrs][index] == [""] || @edit[:new][:attrs][index] == ["", ""]) && @edit[:current][:attrs][index] == [] # check if attribute and value field is empty
```

Comment:
Why are we storing empty strings there?  It makes it hard to check for empty. It should be easier to do `.empty?` but we can't because we have arrays with 1 or 2 empty strings.  It's very strange.

Reply:
As discussed 

```uby
FIELDS_EMPTY = [[''], ['', '']]

...
FIELDS_EMPTY.include?(@edit[:new][:attrs][index])
```